panel.py

```diff
@@ -77,50 +82,126 @@
   sock.sendall(json.dumps({ "message": message, "data": data }) + '\r')
 
-# TODO: Make this repeatable so that we can receive additional messages even if we read past a single
-# message (currently we drop `following_text`), and non-blocking so that it can be interspersed with
-# messages from the controls and consequent announcements to the controller. Not sure how to do
-# those things!
-def receive():
-  buffer = ''
-  while True:
-    data = sock.recv(4096) # Arbitrary / maximum length.
-    buffer += data
-    raw_event, carriage_return, following_text = buffer.partition('\r')
-    if raw_event:
-      return json.loads(raw_event)
-
 def announce():
   send("announce", { "controls": controls })
 
 announce()
 
-# Play the game.
-while (True):
-  event = receive()
-  if event['message'] == 'display':
-    display = event['data']['display']
-    print display
 
-    # HACK(jeff): This script shouldn't have to know what's displayed, this is just for purposes of
-    # playing the game via the CLI.
-    if display == 'Nice job!':
-      continue
+# Handle events from controls and the controller.
+
+# Mimic controls changing by the player entering keyboard input.
+# Use a thread to allow blocking reads thereof.
+
+control_queue = []
+def readKeyboard():
+  while (True):
+    control_id, space, state = raw_input().partition(' ')
+    if control_id and state:
+      control_queue.append((control_id, state))
+threading.Thread(target=readKeyboard).start()
 
-  control_id, space, state = raw_input('Set state: ').partition(' ')
+def receiveControl():
+  try:
+    state_change = control_queue.pop(0) # Pops from start of list.
+  except:
+    # Ignore the queue being empty.
+    return False
+
+  control_id, state = state_change
+
+  try:
+    control = filter(lambda c: c['id'] == control_id, controls)[0]
+  except:
+    # The player entered invalid input like trying to manipulate a control that wasn't on their board.
+    return False
 
   send('set-state', { 'id': control_id, 'state': state })
 
   # HACK(jeff): We need to explicitly reset buttons' state while playing from the CLI, whereas a
   # physical button's state would automatically reset when the player lifted their finger off.
-  control = filter(lambda c: c['id'] == control_id, controls)[0]
   if ('type' in control) and (control['type'] == 'button'):
     send('set-state', { 'id': control_id, 'state': '0' })
 
+  return True
+
+buffer = ''
+def receiveController():
+  global buffer
+
+  while True:
+    try:
+      data = sock.recv(4096) # Arbitrary / maximum length.
+    except:
+      # If the buffer is empty, we return as not having handled an event.
+      # Otherwise we wait for the rest of the data corresponding to this event.
+      if not buffer:
+        return False
+      else:
+        continue
+
+    buffer += data
+
+    # Note that we may receive multiple events at once.
+    while buffer:
+      raw_event, carriage_return, following_text = buffer.partition('\r')
```

Comment:
i don't think we should partition on `'\r'` -- can we use the same encoding scheme as we do in the serial comms (cobs) ? i have a similar data-read loop here: https://github.com/igor47/spaceboard/blob/master/spaceteam/microcontroller.py#L108-L132 . with cobs we'd partition on `'\x00'` and then call `cobs.decode(raw_event)` to get the json

Reply:
otherwise we might get confused if the json includes a carriage return